### PR TITLE
Add exllama GPTQ CUDA kernel support

### PR DESCRIPTION
**Pull request by fxmarty**
Examples:

```
text-generation-launcher --model-id Narsil/starcoder-gptq --port 8080 --num-shard 2 --quantize gptq
GPTQ_GROUPSIZE=128 GPTQ_BITS=4 text-generation-launcher --model-id TheBloke/WizardLM-7B-uncensored-GPTQ --port 8080 --num-shard 2 --quantize gptq
pytest integration-tests/models/test_flash_llama_gptq.py -s
pytest integration-tests/models/test_flash_starcoder_gptq.py -s
```

This PR adds to TGI the mixed precision int4/fp16 kernels from the excellent [exllama repo](https://github.com/turboderp/exllama), that from [my benchmark](https://github.com/fxmarty/q4f16-gemm-gemv-benchmark) is much better than the implementations available in autogptq & gptq-for-llama.

On batch size 1, for starcoder with starcoder & GPTQ-4bit-no-actorder, we get a x2.1 speedup on the prefill over GPTQ-triton, and x1.8 speedup on the decode over GPTQ-triton. I'll have a look at the peak memory.

I verified locally that logits match.

Note that exllama implementation can not be used with act-order & tp rank>=2 for row tensor parallel linear, because exllama reorders weights ahead of runtime, requiring to reorder the activation as well (which are split on several GPUs for row parallel + TP rank>=2). In this specific case, we default to the trition implementation (that is much slower because reordering is done one the scales/zero points, and each weight row need to have its own specific scale/zero point).

Exllama implementation is specifically for n_bits = 4. Thus, for the other cases we fall back on the triton kernel.

Results on starcoder are as follow (TP rank = 2, A100, before vllm):

| Parameter          | Value                 |
|--------------------|-----------------------|
| Model              | Narsil/starcoder-gptq |
| Sequence Length    | 512                   |
| Decode Length      | 512                   |
| N Runs             | 10                    |
| Warmups            | 1                     |
| Temperature        | None                  |
| Top K              | None                  |
| Top P              | None                  |
| Typical P          | None                  |
| Repetition Penalty | None                  |
| Watermark          | false                 |
| Do Sample          | false                 |

<details>
  <summary>GPTQ (current):</summary>
  
| Step           | Batch Size | Average     | Lowest      | Highest     | p50         | p90         | p99         |
|----------------|------------|-------------|-------------|-------------|-------------|-------------|-------------|
| Prefill        | 1          | 139.00 ms   | 138.31 ms   | 139.93 ms   | 139.06 ms   | 139.93 ms   | 139.93 ms   |
|                | 4          | 522.17 ms   | 521.71 ms   | 522.57 ms   | 522.16 ms   | 522.57 ms   | 522.57 ms   |
|                | 8          | 1016.23 ms  | 1015.78 ms  | 1016.87 ms  | 1016.10 ms  | 1016.87 ms  | 1016.87 ms  |
| Decode (token) | 1          | 36.19 ms    | 36.07 ms    | 36.54 ms    | 36.11 ms    | 36.54 ms    | 36.54 ms    |
|                | 4          | 36.65 ms    | 36.57 ms    | 36.74 ms    | 36.64 ms    | 36.74 ms    | 36.74 ms    |
|                | 8          | 36.72 ms    | 36.55 ms    | 36.94 ms    | 36.72 ms    | 36.94 ms    | 36.94 ms    |
| Decode (total) | 1          | 18491.31 ms | 18433.56 ms | 18671.24 ms | 18451.02 ms | 18671.24 ms | 18671.24 ms |
|                | 4          | 18728.41 ms | 18689.53 ms | 18776.16 ms | 18724.24 ms | 18776.16 ms | 18776.16 ms |
|                | 8          | 18762.84 ms | 18678.29 ms | 18875.27 ms | 18763.06 ms | 18875.27 ms | 18875.27 ms |

| Step    | Batch Size | Average           | Lowest            | Highest           |
|---------|------------|-------------------|-------------------|-------------------|
| Prefill | 1          | 7.19 tokens/secs  | 7.15 tokens/secs  | 7.23 tokens/secs  |
|         | 4          | 7.66 tokens/secs   | 7.65 tokens/secs   | 7.67 tokens/secs   |
|         | 8          | 7.87 tokens/secs   | 7.87 tokens/secs   | 7.88 tokens/secs   |
| Decode  | 1          | 27.64 tokens/secs | 27.37 tokens/secs | 27.72 tokens/secs |
|         | 4          | 109.14 tokens/secs | 108.86 tokens/secs | 109.37 tokens/secs |
|         | 8          | 217.88 tokens/secs | 216.58 tokens/secs | 218.86 tokens/secs |
  
</details>

<details>
  <summary>GPTQ-CUDA (exllama):</summary>
  
| Step           | Batch Size | Average     | Lowest     | Highest     | p50        | p90         | p99         |
|----------------|------------|-------------|------------|-------------|------------|-------------|-------------|
| Prefill        | 1          | 65.49 ms    | 64.76 ms   | 65.81 ms    | 65.58 ms   | 65.81 ms    | 65.81 ms    |
|                | 4          | 190.36 ms   | 189.25 ms   | 194.22 ms   | 190.25 ms   | 194.22 ms   | 194.22 ms   |
|                | 8          | 350.45 ms   | 349.56 ms   | 353.83 ms   | 350.07 ms   | 353.83 ms   | 353.83 ms   |
| Decode (token) | 1          | 19.69 ms    | 18.90 ms   | 21.20 ms    | 18.99 ms   | 21.20 ms    | 21.20 ms    |
|                | 4          | 30.51 ms    | 30.42 ms    | 30.64 ms    | 30.49 ms    | 30.58 ms    | 30.58 ms    |
|                | 8          | 34.73 ms    | 34.56 ms    | 34.80 ms    | 34.76 ms    | 34.80 ms    | 34.80 ms    |
| Decode (total) | 1          | 10061.35 ms | 9659.25 ms | 10835.27 ms | 9705.76 ms | 10835.27 ms | 10835.27 ms |
|                | 4          | 15592.28 ms | 15547.03 ms | 15659.89 ms | 15582.29 ms | 15626.84 ms | 15626.84 ms |
|                | 8          | 17749.09 ms | 17661.60 ms | 17781.76 ms | 17760.34 ms | 17781.76 ms | 17781.76 ms |

| Step    | Batch Size | Average           | Lowest            | Highest           |
|---------|------------|-------------------|-------------------|-------------------|
| Prefill | 1          | 15.27 tokens/secs | 15.20 tokens/secs | 15.44 tokens/secs |
|         | 4          | 21.01 tokens/secs  | 20.59 tokens/secs  | 21.14 tokens/secs  |
|         | 8          | 22.83 tokens/secs  | 22.61 tokens/secs  | 22.89 tokens/secs  |
| Decode  | 1          | 50.91 tokens/secs | 47.16 tokens/secs | 52.90 tokens/secs |
|         | 4          | 131.09 tokens/secs | 130.52 tokens/secs | 131.47 tokens/secs |
|         | 8          | 230.32 tokens/secs | 229.90 tokens/secs | 231.46 tokens/secs |

</details>

## Before submitting
- [ ] Doc ==> There's no doc currently right?
- [x] Tests ==> Done for starcoder and llama

### CODE REVIEW

**Review by Narsil — COMMENTED** [review]
Neat numbers ! 

I feel like `gptq` and `gptq-cuda` is not necessary here.

IIUC, both can run on the same weights (as you didn't change the conversion script).
Therefore, we could simply use `exllama` kernel whenever available (when g_idx is increasing).

That should simplify the codebase a lot.

Also nothing should be modified in model files, everything should be very agnostic to it, especially since the weights are exactly the same on disk.

This could also be explained in the gptq script  (act-order, if True, more precise models, but slower inference because different kernels, if False, lower precision, but ffaster inference).

I still fail to understand why we cannot reorder on load to use exllama for act-order (since we can reslice at will in the original tensors, we could probably de-entagle `g_idx` again.

It's a lot more work certainly.

File: server/text_generation_server/models/custom_modeling/flash_santacoder_modeling.py [hunk 1]
```diff
@@ -179,22 +181,40 @@ def load_col(config, prefix: str, weights, bias: bool):
...
+        weight = weights.get_multi_weights_row(prefix, quantize=quantize)
 
     if bias and weights.process_group.rank() == 0:
         # Rank is only on the first rank process
         bias = weights.get_tensor(f"{prefix}.bias")
     else:
         bias = None
+    
+    if quantize == "gptq-cuda" and not act_order:
+        weight[3] = None  # remove g_idx to indicate to exllama that act-order is not used
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/flash_santacoder_modeling.py:214]: Just don't modify anything here, let this flow to `get_linear` which can do;

```python
if torch.equal(g_idx.cpu(), torch.tensor([i // groupsize for i in range(g_idx.shape[0])], dtype=torch.int32)) or (g_idx == 0).all():
```

and decide there if we should use the triton or the cuda kernel.

File: server/text_generation_server/models/custom_modeling/flash_santacoder_modeling.py [hunk 2]
```diff
@@ -456,6 +478,30 @@ def __init__(self, config, weights):
...
+            prepare_buffers(weights.device, self.buffers["temp_state"], self.buffers["temp_dq"])
+
+            # TODO: ability to set them
+            matmul_recons_thd = 8
+            matmul_fused_remap = False
+            matmul_no_half2 = False
+            set_tuning_params(matmul_recons_thd, matmul_fused_remap, matmul_no_half2)
+
+            torch.cuda.empty_cache()
+
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply server/text_generation_server/models/custom_modeling/flash_santacoder_modeling.py]: I moved it to `Model` init. This requires to have access to `model.config` which is currently not defined though. There is `model.transformer.config`, or `model.gpt_neox.config` or `model.model.config` depending on the architecture. Is it intended that the config is not registered at the top level? @OlivierDehaene @Narsil 

The thing is that the `weights = Weights(...)` call is in each model definition, and we need to have loaded all weights to determine the shapes of the buffers. Also, the buffers need to be persistent, while I think this `weights` object is not.

[the remark this answers is not in the document]
Comment (Narsil) [reply server/text_generation_server/models/custom_modeling/flash_santacoder_modeling.py]: The buffers intend to be shared, no ?

So why not just have a single location for these buffers, use the pointer on every layer, and increase the size every time `max_dq_buffer_size = max(max_dq_buffer_size, submodule.linear.qweight.numel() * 8)` is larger ?

The issue with this (and any post loading treatment) is that you're now dealing with updating every single model file, any time of those line hits. This is what we had before and it was painful to maintain.

This seems to be used as globals let's just use them as globals. (They are temporary buffers IIUC preallocated to avoid reallocating them all the time)

**Comment by fxmarty** [comment]
For some reason the `logprob` slightly differ between different runs, there's a source of randomicity I've still yet not identified.

**Comment by Narsil** [comment]
> atomicAdd + randomicity

This is very suspicious, really ? Isn't the purpose of atomicAdd to remove randomness by forcing access order ? :)
It may well be acceptable though :)

**Comment by fxmarty** [comment]
@Narsil I don't believe it is suspicious: https://forums.developer.nvidia.com/t/get-different-results-for-every-running-with-atomicadd/229649/2

See https://github.com/turboderp/exllama/issues/153

**Comment by fxmarty** [comment]
Some tests up to date with main on llama 2 70b

image

**Review by Narsil — APPROVED** [review]
